models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,486 @@
+
+nest::gif_psc_exp::gif_psc_exp()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::gif_psc_exp::gif_psc_exp( const gif_psc_exp& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::gif_psc_exp::init_state_( const Node& proto )
+{
+  const gif_psc_exp& pr = downcast< gif_psc_exp >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::gif_psc_exp::init_buffers_()
+{
+  B_.spikes_ex_.clear(); // includes resize
+  B_.spikes_in_.clear(); // includes resize
+  B_.currents_.clear();  //!< includes resize
+  B_.logger_.reset();    //!< includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::gif_psc_exp::calibrate()
+{
+
+  B_.logger_.init();
+
+  double_t h = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.P11ex_ = std::exp( -h / P_.tau_ex_ );
+  V_.P11in_ = std::exp( -h / P_.tau_in_ );
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, h );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, h );
+
+
+  V_.P33_ = std::exp( -h / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_
+    >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.sfa_stc_initialized_ )
+  {
+    V_.P_sfa_.clear();
+    V_.P_stc_.clear();
+    S_.sfa_elems_.clear();
+    S_.stc_elems_.clear();
+
+    for ( size_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.P_sfa_.push_back( std::exp( -h / P_.tau_sfa_[ i ] ) );
+      S_.sfa_elems_.push_back( 0.0 );
+    }
+
+    for ( size_t i = 0; i < P_.tau_stc_.size(); i++ )
+    {
+      V_.P_stc_.push_back( std::exp( -h / P_.tau_stc_[ i ] ) );
+      S_.stc_elems_.push_back( 0.0 );
+    }
+
+    S_.sfa_stc_initialized_ = true;
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ */
+
+void
+nest::gif_psc_exp::update( Time const& origin,
+  const long_t from,
+  const long_t to )
+{
+
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  double_t q_temp_;
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    // exponential decaying stc and sfa elements
+    q_temp_ = 0;
+    for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.stc_elems_[ i ];
+
+      S_.stc_elems_[ i ] = V_.P_stc_[ i ] * S_.stc_elems_[ i ];
+    }
+
+    S_.stc_ = q_temp_;
+
+    q_temp_ = 0;
```

Comment:
Thanks for the suggestions. 
For the first suggestion, I did not use `S_.stc_` for the performance. Because each time the compiler needs to modify the value of `S_.stc_` it has to follow two pointers to reach the variable (one for the struct and another one for the member of the struct). By using `q_temp_` I reduced it to following only one pointer. Actually, the whole idea of using `q_temp_` is due to the performance. Otherwise, the could would look more readable and beautiful using the members of `S_` directly. 
For the second suggestion, I had the same idea. However, since I am not sure about implementation of loggers in the NEST, I was afraid of implementing that. What does happen if two voltmeters connect to one neuron? Does each of them call `get_stc_()` separately? If so, the stc elements decay twice in each time step.   

I have changed `q_temp_` to `q_temp`.